### DISCUSSION TITLE
flash_map_default.c.obj): undefined reference to `__device_dts_ord_xx' when I convert my SPI flash driver into a out-of-tree driver.

### DISCUSSION THREAD

**Discussion by gineelee** [zephyrproject-rtos/zephyr#57764] [Q&A]
Everyting works find when I move my spi flash driver out of Zephyr codebase to be an out-of-tree driver.
The flash driver has been built as an external module called 'externalLib'

Then I added the partitions in the device tree as below since I want to utilize Zephyr's FS subsystem on my external flash.
` &mx25rxxx {
	partitions {
		compatible = "fixed-partitions";
		#address-cells = <1>;
		#size-cells = <1>;

		littlefs_dev: partition@0 {
			label = "littlefs_dev";
			reg = <0x00000000 0x00010000>;
		};
	};
};`

I can see the partitions of the spi flash are created correctly in devicetree_generated.h. (PARTITION_ID 0~3 are allocated for internal flash.)
`/* fixed-partitions identifier: */
#define DT_N_S_soc_S_octospi_420d1400_S_ospi_flash_0_S_partitions_S_partition_0_PARTITION_ID 4`

However, there is a error when bulding mcuboot.
`[1/9] Linking C executable zephyr/zephyr_pre0.elf
FAILED:ld.bfd: zephyr/libzephyr.a(flash_map_default.c.obj):(.rodata.default_flash_map+0x4c): undefined reference to __device_dts_ord_98`

Upon checking the zephyr.map file, I noticed that '__device_dts_ord_98' corresponds to my spi flash driver. However, it seems that mcuboot is unable to reference my external module 'externalLib,' which includes the spi flash driver. Interestingly, there is no mention of '__device_dts_ord_98' in mcuboot's 'zephyr_pre0' file.

It seems strange to me since both mcuboot and externalLib are external modules for my main application."
My question is whether it's possible for mcuboot to include my additional external module 'externalLib' during the build process.

Thanks!

**Comment by gineelee** [comment]
It is how I mount the file system in application.
```c
#define STORAGE_PARTITION	        littlefs_dev
#define STORAGE_PARTITION_ID	FIXED_PARTITION_ID(STORAGE_PARTITION)

static struct fs_mount_t lfs_mnt = {
    .type = FS_LITTLEFS,
    .fs_data = &cfg,
    .storage_dev = (void*)STORAGE_PARTITION_ID,
    .mnt_point = "/littlefs",
};
...
fs_mount(&lfs_mnt);
...
```

**Comment by Antatip** [comment]
Hi,

I am having the same trouble building my project.
As soon as I enable the `CONFIG_BOOTLOADER_MCUBOOT` flag build fails.
I get the log `=== child image mcuboot -  end ===` so I guess this is my app which is not compiling well.

The `__device_dts_ord_xx` error is clearly my external flash which is not found.

This [post](https://devzone.nordicsemi.com/f/nordic-q-a/109144/recognizing-out-of-tree-device-driver-bindings-for-mcuboot-child-image) says out of tree driver is not known by MCUBOOT build but as I say it seems to be built.
I also tried the `CONFIG_PM_OVERRIDE_EXTERNAL_DRIVER_CHECK` flag without success.

Here is the interesting part of my dts file : 
```
/dts-v1/;
#include <nordic/nrf52832_qfaa.dtsi>

/ {

	chosen {
		zephyr,flash = &flash0;
		nordic,pm-ext-flash = &ext_flash;
	};

	aliases {
		ext-flash = &ext_flash;
	};
};

&flash0 {
	partitions {
		compatible = "fixed-partitions";
		#address-cells = <1>;
		#size-cells = <1>;

		boot_partition: partition@0 {
			label = "mcuboot";
			reg = <0x00000000 0xC000>;
		};
		slot0_partition: partition@C000 {
			label = "image-0";
			reg = <0x0000C000 0x74000>;
		};
	};
};

&spi2 {
	status = "okay";
	pinctrl-0 = <&spi2_default>;
	pinctrl-names = "default";
	cs-gpios = <&gpio0 29 GPIO_ACTIVE_LOW>,
                <&gpio0 5 GPIO_ACTIVE_LOW>;
	ext_flash: ext_flash@0{
		compatible = "jedec,spi-nor";
		spi-max-frequency = < DT_FREQ_M(50) >;
		reg = <0>;
		hold-gpios = <&gpio0 8 (GPIO_ACTIVE_LOW)>;
		wp-gpios = <&gpio0 7 (GPIO_ACTIVE_LOW)>;
		jedec-id = [ BF 25 8D ];
		size = <DT_SIZE_M(4*8)>;
		has-lock = < 0xBC >;
		partitions{
			compatible = "fixed-partitions";
			#address-cells = <1>;
			#size-cells = <1>;
	
			slot1_partition: partition@0 {
				label = "image-1";
				reg = <0x00000000 0x000074000>;
			};
			user_partition: partition@74000 {
				label = "storage";
				reg = <0x000074000 0x0001F8C000>;
			};
		};
	};
};
```

Anyone knowing what I am doing wrong ?

**Reply by nordicjm** [reply]
You are using ncs, not zephyr, you need a static PM file

**Reply by Antatip** [reply]
I do too but I am not very sure of its content : 
```
app:
  address: 0xC000
  region: flash_primary
  size: 0x74000
mcuboot:
  address: 0x0
  region: flash_primary
  size: 0xC000

external_flash:
  address: 0x00
  end_address: 0x2000000
  region: external_flash
  size: 0x2000000

mcuboot_secondary:
  address: 0x00000
  device: ext_flash
  end_address: 0x74000
  region: external_flash
  size: 0x74000
```

Also, what does it have to do with the driver build ?
